common/src/main/java/com/revenuecat/purchases/common/networking/ETagManager.kt

```diff
@@ -0,0 +1,120 @@
+package com.revenuecat.purchases.common.networking
+
+import android.content.Context
+import android.content.SharedPreferences
+import com.revenuecat.purchases.common.sha1
+import org.json.JSONObject
+import java.net.HttpURLConnection
+
+private const val NOT_MODIFIED_RESPONSE_CODE = 304
+private const val INTERNAL_SERVER_ERROR_RESPONSE_CODE = 500
+private typealias RequestHash = String
+
+private const val SERIALIZATION_NAME_ETAG = "eTag"
+private const val SERIALIZATION_NAME_HTTPRESULT = "httpResult"
+
+internal const val ETAG_HEADER_NAME = "X-RevenueCat-ETag"
+
+data class HTTPResultWithETag(
+    val eTag: String,
+    val httpResult: HTTPResult
+) {
+    fun serialize(): String {
+        return JSONObject().apply {
+            put(SERIALIZATION_NAME_ETAG, eTag)
+            put(SERIALIZATION_NAME_HTTPRESULT, httpResult.serialize())
+        }.toString()
+    }
+
+    companion object {
+        fun deserialize(serialized: String): HTTPResultWithETag {
+            val jsonObject = JSONObject(serialized)
+            val eTag = jsonObject.getString(SERIALIZATION_NAME_ETAG)
+            val serializedHTTPResult = jsonObject.getString(SERIALIZATION_NAME_HTTPRESULT)
+            return HTTPResultWithETag(eTag, HTTPResult.deserialize(serializedHTTPResult))
+        }
+    }
+}
+
+class ETagManager(
+    private val prefs: SharedPreferences
+) {
+    private var hashesByHttpRequest: Map<HTTPRequest, RequestHash> = emptyMap()
```

Comment:
@aboedo @beylmk When I was developing the iOS version I realized that we don't have to use hashes of the URL and we can just use the URL. I initially was hashing a combination of URL + body, but we moved away from that and now we are only using the URL to identify the etags.

Do you think I should remove the use of hashed URLs, or should I leave it as is. Pros of not using hashes is a simpler implementation with less chances of breaking plus less computing time, cons are that the key in SharedPreferences is going to be the full URL, which can get long.

What are your thoughts?

Reply:
I would argue speed/simplicity wins...though maybe it depends on what our "cleanup" strategy was. How often/when did we decide to clear etags sharedprefs? on logout?

I'm also still a little hesitant on the security of storing user IDs. Trying to get over it though 😅

Reply:
I'd agree that simplicity wins. As for the removing the base URL... That'd be neat! The only downside I can think of is that we'd have to remember to clean install the app when switching servers for development (i.e.: running against a local backend). 

Which... we need to do anyway, really, because any existing data wouldn't work with a different server, so I think it'd be fine.